src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,228 @@
+
+#[cfg(test)]
+mod tests;
+
+// TIERED SPARSE MERKLE TREE
+// ================================================================================================
+
+pub struct TieredSmt {
+    // TODO this should be a constant
+    empty_subtrees: Vec<Word>,
+    root: Word,
+    storage: Storage,
+}
+
+impl TieredSmt {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    pub const MAX_DEPTH: usize = 64;
+    pub const TIER_DEPTH: usize = 16;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn with_storage(storage: Storage) -> Result<Self, StorageError> {
+        // Construct empty node digests for each layer of the tree
+        let empty_subtrees: Vec<_> = (0..Self::MAX_DEPTH + 1)
+            .scan(RpoDigest::default(), |state, _| {
+                let value = *state;
+                *state = Rpo256::merge(&[value, value]);
+                Some(value.into())
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
+            .collect();
+
+        let root = Word::default();
+        let mut tree = Self {
+            empty_subtrees,
+            root,
+            storage,
+        };
+        tree.root = tree.get_node(&NodeIndex::root())?;
+        Ok(tree)
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn get_leaf_value(&self, key: &Word) -> Result<Option<Word>, StorageError> {
+        let index = match self.get_leaf_index(key)? {
+            Some(index) => index,
+            None => return Ok(None),
+        };
+        match self.storage.get_leaves(&index)? {
+            Some(l) => Ok(l
+                .into_iter()
+                .find_map(|leaf| (&leaf.key == key).then_some(leaf.value))),
+            None => Ok(None),
+        }
+    }
+
+    pub const fn root(&self) -> &Word {
+        &self.root
+    }
+
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn get_path(&self, key: &Word) -> Result<Option<TieredMerklePath>, StorageError> {
+        let leaf_index = match self.get_leaf_index(key)? {
+            Some(index) => index,
+            None => return Ok(None),
+        };
+        let leaves = self
+            .storage
+            .get_leaves(&leaf_index)?
+            .ok_or(StorageError::LeafNodeWithoutDigest)?;
+        let mut path = Vec::with_capacity(leaf_index.depth() as usize);
+        let mut index = leaf_index;
+        while index.depth() > 0 {
+            let sibling = self.get_node(&index.sibling())?;
+            path.push(sibling);
+            index = index.reverse();
+        }
+        let path = MerklePath::new(path);
+        let path = TieredMerklePath::new(leaf_index, leaves, path);
+        Ok(Some(path))
+    }
```

Comment:
Shouldn't this always return a path? Meaning, shouldn't the signature be something like:
```Rust
pub fn get_path(&self, key: &Word) -> Result<MerklePath, StorageError>
```
The idea is that even if the tree does not have a leaf at this key, we should return a path to an empty node. This way, we can also have exclusion proofs. There is probably some extra handling needed for values at depth 64 - but we can get to that later.

Also, I'm not sure we need to return anything more complex than a `MerklePath`. We should be able to infer the index from the combination of `key` and the length of the returned path. We also don't need to include leaf info with the path.

Reply:
This is also fixed. This function is closer to a `get_leaf_path` to a `get_node_path`. The former will return `None` if the leaf doesn't exist in the tree, and the latter will always return a path.

The latter is now available as internal helper.

Reply:
I would probably change this and make `get_path()` to be basically `get_node_path()` but exposed publicly. We may also be able to get away with just this method instead of having `prove_membership()` and `prove_non_membership()` - but I'll need to think about this a bit more.